main.go

```diff
@@ -1,196 +1,861 @@
+	goType = fd.GetTypeName()
+	if i := strings.LastIndex(goType, "."); i > 0 {
+		goType = goType[i+1:]
 	}
-	fmt.Fprintf(w, "	return []string{%s}\n", strings.Join(wrappedPaths, ", "))
-	fmt.Fprintln(w, "}")
-}
 
-func (p *plugin) registerMessage(messages map[string]*protokit.Descriptor, md *protokit.Descriptor) {
-	for _, sub := range md.GetMessages() {
-		if strings.HasSuffix(sub.GetName(), "Entry") {
-			continue
+	protoType := fd.GetTypeName()[1:]
+
+	for parent := fd.GetMessage(); parent != nil; parent = parent.GetParent() {
+		for _, sed := range parent.GetEnums() {
+			if protoType == sed.GetFullName() {
+				goType = fmt.Sprintf("%s_%s", parent.GetName(), goType)
+				if i := strings.LastIndex(goType, "."); i > 0 {
+					goType = goType[i+1:]
+				}
+				protoType = parent.GetFullName()
+			}
 		}
-		sub.Name = proto.String(fmt.Sprintf("%s_%s", md.GetName(), sub.GetName()))
-		p.registerMessage(messages, sub)
 	}
-	messages[md.FullName] = md
+	return goType
 }
 
-func (p *plugin) appendDescriptor(s []*protokit.Descriptor, md *protokit.Descriptor) []*protokit.Descriptor {
-	for _, mfd := range md.GetMessageFields() {
-		if mfd.GetTypeName() == ".google.protobuf.FieldMask" {
-			return s
-		}
+var importPathReplacer = strings.NewReplacer(
+	".", "_",
+	"/", "_",
+	"-", "_",
+)
+
+func buildMethods(buf *strings.Builder, md *protokit.Descriptor, mdMap map[string]*protokit.Descriptor) (map[string]string, error) {
+	imports := map[string]string{
+		"types": "github.com/gogo/protobuf/types",
 	}
-	out := append(s, md)
-	for _, sub := range md.GetMessages() {
-		if strings.HasSuffix(sub.GetName(), "Entry") {
-			continue
-		}
-		out = p.appendDescriptor(out, sub)
+
+	paths, err := appendPaths(make([]string, 0, len(md.GetMessageFields())), "", md, mdMap, nil)
+	if err != nil {
+		// TODO: Return error here
+		log.Printf("Failed to traverse `%s`: %s, skipping...", md.GetFullName(), err)
+		return nil, nil
 	}
-	return out
+
+	fmt.Fprintf(buf, `
+var _%sFieldPaths = [...]string{%s}
+
+func (*%s) FieldMaskPaths() []string {`,
+		md.GetName(), `"`+strings.Join(paths, `", "`)+`"`,
+		md.GetName(),
+	)
+
+	sort.Strings(paths)
+	if len(paths) == 0 {
+		fmt.Fprintf(buf, `
+	return nil`,
+		)
+	} else {
+		fmt.Fprintf(buf, `
+	ret := make([]string, len(_%sFieldPaths))
+	copy(ret, _%sFieldPaths[:])
+	return ret`,
+			md.GetName(),
+			md.GetName(),
+		)
+	}
+
+	fmt.Fprintf(buf, `
 }
 
-func (p *plugin) Generate(in *plugin_go.CodeGeneratorRequest) (*plugin_go.CodeGeneratorResponse, error) {
-	descriptors := protokit.ParseCodeGenRequest(in)
+func (dst *%s) SetFields(src *%s, mask *types.FieldMask) {
+	for _, path := range mask.GetPaths() {
+		switch path {`,
+		md.GetName(), md.GetName(),
+	)
 
-	messages := make(map[string]*protokit.Descriptor)
-	for _, fd := range descriptors {
-		for _, md := range fd.GetMessages() {
-			p.registerMessage(messages, md)
+	for _, p := range paths {
+		fmt.Fprintf(buf, `
+		case "%s":
+			var nilPath bool`,
```

Comment:
the variable `nilPath` is set only in L520 for OneOfs. I assume that the go complier is smart enough to remove this during optimization but it's redundant for non-oneof fields.

Reply:
`nilPath` is not set on `L520` - it's only evaluated there, it is set at `L258` and `L277`, however.
As you can see, it not redundant for non-oneof fields.
It is only redundant for field paths, which do not contain pointers in any of the links.

Reply:
Because if we're setting `A.B.C` we don't care if `A.D` is `nil`
`nilPath` only concerns the intermidiary steps in the path(so for `A.B.C.D` `nilPath == true` implies that one of `{,A,A.B,A.B.C}` is `nil`)

Reply:
I also agree with @htdvisser. Since the scope of the variable is only within the the case, in the above snippet it's redundant  and the `if nilpath {}` is dead code that I assume that the go compiler will remove. You could define this variable above the switch and reset upon entering a branch.